File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {number}
 */
var pairSum = function(head) {

    const arr = [];
    let max = 0;

    let stack = [head];
    arr.push(head.val)

    while(stack.length) {
        let current = stack.pop()
        if (current.next !== null) {
            stack.push(current.next)
            arr.push(current.next.val)
        }
    }

    for (let i = 0; i < arr.length / 2; i++) {
        if (arr[i] + arr[arr.length-1-i] > max) {
            max = arr[i] + arr[arr.length-1-i]
        }
    }

    return max
    
};


/**
Rank 199,606
Date: 05/16/2023
Runtime: 181 ms, faster than 43.54%
Memory Usage: 91.9 MB, less than 23.25% 
 */

File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        
        # approach 1: convert into a list
        # approach 2: mutate nodes to have a prev pointer
        # approach 3: fast and slow pointer to find the middle of linked list
        #             then reverse the second half of linked list, then traverse both 
        
        slow, fast = head, head
        max_sum = 0
        prev = None
        
        # find the middle of the linked list
        # when fast is out of bound, slow is second half of linked list
        # since linked list is even, we don't need to check if fast.next is NOne
        while fast:
            slow = slow.next
            fast = fast.next.next
        
        #reverse second half of the linked list
        while slow:
            temp = slow.next
            slow.next = prev
            prev = slow
            slow = temp
            
        #run two linked list together
        while prev:
            twin_sum = head.val + prev.val
            max_sum = max(max_sum, twin_sum)
            
            prev = prev.next
            head = head.next
            
        return max_sum
        